continuationsPlugin/src/main/scala/continuations/BodyHasNoSuspensionPoint.scala

```diff
@@ -0,0 +1,32 @@
+package continuations
+
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.tpd.*
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Symbols
+
+/**
+ * A matcher for identifying methods with suspending continuations in their bodies.
+ */
+private[continuations] object BodyHasNoSuspensionPoint {
+
+  /**
+   * Matches methods with a using [continuations.Suspend] parameter that suspend a continuation
+   * in their method body.
+   *
+   * @param defdef
+   *   The method to match upon
+   * @return
+   *   A [scala.Some] if the method has a using [continuations.Suspend] parameter and has a
+   *   suspended continuation in its method body, [scala.None] otherwise
+   */
+  def unapply(defdef: tpd.DefDef)(using Context): Option[tpd.DefDef] =
+    val suspendType = Symbols.requiredMethod("continuations.Continuation.suspendContinuation")
+    Option(defdef).filter {
+      case HasSuspendParameter(_) =>
+        defdef.rhs.existsSubTree{ subtree =>
```

Comment:
depending on whether we have `suspendContinuation` inlined or not this may not work.

if it is inlined I think the subtrees will traverse the code that is inlined which will be`???` instead of the to-be-inlined one, hence will need something like [this](https://github.com/47deg/TBD/pull/52/files#diff-fade65ebf959155389570d62d274801c2f7607f965a273661e5f776636d6393fR164)

Reply:
Good catch. We need to add a test case using `Inlined`, certainly.

I think the `existsSubTree` uses the `foldover` just like `filterSubTree`:
 
```scala
    def foreachSubTree(f: Tree => Unit)(using Context): Unit = {
      val traverser = new TreeTraverser {
        def traverse(tree: Tree)(using Context) = foldOver(f(tree), tree)
      }
      traverser.traverse(tree)
    }


    /** Is there a subtree of this tree that satisfies predicate `p`? */
    def existsSubTree(p: Tree => Boolean)(using Context): Boolean = {
      val acc = new TreeAccumulator[Boolean] {
        def apply(x: Boolean, t: Tree)(using Context) = x || p(t) || foldOver(x, t)
      }
      acc(false, tree)
    }


    /** All subtrees of this tree that satisfy predicate `p`. */
    def filterSubTrees(f: Tree => Boolean)(using Context): List[Tree] = {
      val buf = mutable.ListBuffer[Tree]()
      foreachSubTree { tree => if (f(tree)) buf += tree }
      buf.toList
    }
```
[Source](https://github.com/lampepfl/dotty/blob/63344e7337727f886d3f7f1a0ad0234dd7670df8/compiler/src/dotty/tools/dotc/ast/tpd.scala#L1102-L1122)

so I don't think there's anything particular about the `Inlined` tree for which the two techniques wouldn't work essentially the same.

They both delegate to `TreeAccumulator` and traverse into each subtree of each tree one at a time in `foldOver`, recursively so should eventually hit the deconstructed `Inlined[T]#call` tree [here](https://github.com/lampepfl/dotty/blob/d8a6751a8d10c73e5a09245cf1456fce4861763b/tests/pos-with-compiler-cc/dotc/ast/Trees.scala#L1583-L1584) and do the same thing as our `filterSubTrees` predicate linked above. We absolutely do need to test it though, because I could be reading the source incorrectly.

Reply:
@gregorath -- you were correct. Addressed in https://github.com/47deg/TBD/pull/53/commits/a4ccd1d7a19ecfd29c3d2e627183f9a77957d671

Please resolve the conversation if you agree the problem is addressed.

Reply:
yeah it is a bit annoying for our case but the `TreeAccumulator` works with the result code of the Inlined. which makes sense I guess, I did try to see if there was a way to say "for now use only the initial code" or if depending on the phase we could get different behaviour but no luck